File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/arch/mvp/SimplePresenter.java
package com.github.klee0kai.hummus.arch.mvp;

import java.util.WeakHashMap;

public class SimplePresenter implements ISimplePresenter {

    protected final WeakViewsList views = new WeakViewsList();
    protected final WeakHashMap<String, Object> viewsStates = new WeakHashMap<>();

    @Override
    public Object getState(String stateId) {
        return viewsStates.get(stateId);
    }

    @Override
    public void saveState(String stateId, Object state) {
        viewsStates.put(stateId, state);
    }


    @Override
    public void subscribe(IRefreshView view) {
        views.add(view);
    }

    @Override
    public void unsubscribe(IRefreshView view) {
        views.remove(view);
    }

}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/arch/mvp/SimplePresenter.kt
package com.github.klee0kai.hummus.arch.mvp

import java.util.*

open class SimplePresenter : ISimplePresenter {

    protected val views = WeakViewsList()
    protected val viewsStates = WeakHashMap<String?, Any?>()

    override fun getState(stateId: String?): Any? {
        return viewsStates[stateId]
    }

    override fun saveState(stateId: String?, state: Any?) {
        viewsStates[stateId] = state
    }

    override fun subscribe(view: IRefreshView?) {
        views.add(view)
    }

    override fun unsubscribe(view: IRefreshView?) {
        views.remove(view)
    }
}